apps/ejabberd/src/mongoose_commands.erl

```diff
@@ -0,0 +1,567 @@
+
+-spec result(mongoose_command()) -> term().
+result(Cmd) ->
+    Cmd#mongoose_command.result.
+
+%% @doc Command execution.
+-spec execute(caller(), atom()|mongoose_command(), [term()]|map()) -> {ok, term()} | ok | failure().
+execute(Caller, Name, Args) when is_atom(Name) ->
+    case ets:lookup(mongoose_commands, Name) of
+        [Command] -> execute_command(Caller, Command, Args);
+        [] -> {error, not_implemented, <<"This command is not supported">>}
+    end;
+execute(Caller, #mongoose_command{name = Name}, Args) ->
+    execute(Caller, Name, Args).
+
+init() ->
+    case ets:info(mongoose_commands) of
+        undefined ->
+            ets:new(mongoose_commands, [named_table, set, public,
+                {keypos, #mongoose_command.name}]);
+        _ ->
+            ok
+    end.
+
+%%%% end of API
+
+-spec register_commands([mongoose_command()]) -> ok.
+register_commands(Commands) ->
+    lists:foreach(
+        fun(Command) ->
+            check_registration(Command), %% may throw
+            ets:insert_new(mongoose_commands, Command),
+            ejabberd_hooks:run(register_command, [Command]),
+            ok
+        end,
+        Commands).
+
+
+-spec unregister_commands([mongoose_command()]) -> ok.
+unregister_commands(Commands) ->
+    lists:foreach(
+        fun(Command) ->
+            ets:delete_object(mongoose_commands, Command),
+            ejabberd_hooks:run(unregister_command, [Command])
+        end,
+        Commands).
+
+execute_command(Caller, Command, Args) ->
+    try check_and_execute(Caller, Command, Args) of
+        ignore_result ->
+            ok;
+        Res ->
+            {ok, Res}
+    catch
+        {type_mismatch, E} ->
+            {error, type_error, E};
+        permission_denied ->
+            {error, denied, <<"Command not available for this user">>};
+        caller_jid_mismatch ->
+            {error, denied, <<"Caller ids do not match">>};
+        X:E ->
+            ?ERROR_MSG("Caught ~p:~p while executing ~p", [X, E, Command#mongoose_command.name]),
+            {error, internal, term_to_binary(E)}
+    end.
+
+%% @doc This performs many checks - types, permissions etc, may throw one of many exceptions
+%% returns what the func returned or just ok if command spec tells so
+-spec check_and_execute(caller(), mongoose_command(), [term()]) -> term().
+check_and_execute(Caller, Command, Args) when is_map(Args) ->
+    check_and_execute(Caller, Command, map_to_list(Args, Command#mongoose_command.args));
+check_and_execute(Caller, Command, Args) ->
+    % check permissions
+    case is_available_for(Caller, Command) of
+        true ->
+            ok;
+        false ->
+            throw(permission_denied)
+    end,
+    % check caller (if it is given in args, and the engine is called by a 'real' user, then it must match
+    check_caller(Caller, Command, Args),
+    % check args
+    SpecLen = length(Command#mongoose_command.args),
+    ALen = length(Args),
+    if SpecLen =/= ALen ->
```

Comment:
This pretty `ifs` in Erlang :) I think this whole logic will be more readable if moved to a function (or `case` at least).